hypershift-operator/controllers/nodepool/kubevirt/imagecacher.go

```diff
@@ -0,0 +1,249 @@
+			err := cl.Delete(ctx, &oldDV)
+			if err != nil {
+				logger.Error(err, fmt.Sprintf("failed to delete an old DataVolume; namespace: %s, name: %s", oldDV.Namespace, oldDV.Name))
+			}
+		}
+	}
+}
+
+func (qi *cachedQCOWImage) createDVForCache(ctx context.Context, cl client.Client, nodePool *hyperv1.NodePool, infraId string) (*v1beta1.DataVolume, error) {
+	dv := qi.buildDVForCache(nodePool, infraId)
+
+	err := cl.Create(ctx, dv)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return nil, fmt.Errorf("failed to create a DataVolume for the boot image cache ; %w", err)
+	}
+
+	return dv, nil
+}
+
+func (qi *cachedQCOWImage) getDVSourceForVMTemplate() *v1beta1.DataVolumeSource {
+	return &v1beta1.DataVolumeSource{
+		PVC: &v1beta1.DataVolumeSourcePVC{
+			Namespace: qi.namespace,
+			Name:      qi.dvName,
+		},
+	}
+}
+
+func (qi *cachedQCOWImage) buildDVForCache(nodePool *hyperv1.NodePool, infraId string) *v1beta1.DataVolume {
+	dv := &v1beta1.DataVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: bootImageNamePrefix,
+			Namespace:    qi.namespace,
+			Labels: map[string]string{
+				bootImageDVLabelRoleName: bootImageDVLabelRoleValue,
+				bootImageDVLabelInfraId:  infraId,
+			},
+			Annotations: map[string]string{
+				bootImageDVAnnotationHash:          qi.hash,
+				cdiImmediateBindingAnnotation:      "true",
+				cdiDeleteAfterCompletionAnnotation: "false",
+			},
+		},
+		Spec: v1beta1.DataVolumeSpec{
+			Source: &v1beta1.DataVolumeSource{
+				HTTP: &v1beta1.DataVolumeSourceHTTP{
+					URL: qi.name,
+				},
+			},
+			Preallocation: pointer.Bool(true),
+		},
+	}
+
+	kvPlatform := nodePool.Spec.Platform.Kubevirt
+	if kvPlatform.RootVolume != nil && kvPlatform.RootVolume.Persistent != nil {
+		storageSpec := &v1beta1.StorageSpec{}
+		if kvPlatform.RootVolume.Persistent.Size != nil {
+			storageSpec.Resources = corev1.ResourceRequirements{
+				Requests: map[corev1.ResourceName]apiresource.Quantity{
+					corev1.ResourceStorage: *kvPlatform.RootVolume.Persistent.Size,
+				},
+			}
+		}
+
+		if kvPlatform.RootVolume.Persistent.StorageClass != nil {
+			storageSpec.StorageClassName = kvPlatform.RootVolume.Persistent.StorageClass
+		}
+
+		for _, am := range kvPlatform.RootVolume.Persistent.AccessModes {
+			storageSpec.AccessModes = append(storageSpec.AccessModes, corev1.PersistentVolumeAccessMode(am))
+		}
+
+		dv.Spec.Storage = storageSpec
+	}
+
+	return dv
+}
+
+func getCacheDVSelector(infraId string) client.MatchingLabels {
+	return map[string]string{
+		bootImageDVLabelRoleName: bootImageDVLabelRoleValue,
+		bootImageDVLabelInfraId:  infraId,
+	}
+}
+
+func getCacheDVs(ctx context.Context, cl client.Client, infraId string, namespace string) ([]v1beta1.DataVolume, error) {
+	dvs := &v1beta1.DataVolumeList{}
+
+	err := cl.List(ctx, dvs, client.InNamespace(namespace), getCacheDVSelector(infraId))
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to read DataVolumes; %w", err)
+	}
+
+	return dvs.Items, nil
```

Comment:
This `List` is a bit problematic for external infra because we wont' have an informer to pull from.

Is there a way we can make this image a direct lookup with a `Get` instead?

Reply:
I think that the only way to use Get, id with the DataVolume name, but we don't know it because it's random.